File: Algorithm/subjects/101-200/198. 打家劫舍/main.cpp
class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(nums.size(), vector<int>(2));
        // 边界条件
        // dp[i][0] 表示该家没偷
        // dp[i][1] 表示该家偷了
        dp[0][0] = 0;
        dp[0][1] = nums[0];
        for (int i = 1; i < nums.size(); i++) {
            
            dp[i][0] = max(dp[i - 1][0], dp[i - 1][1]);
            dp[i][1] = max(dp[i - 1][1], dp[i - 1][0] + nums[i]);
        }
        return max(dp[n - 1][0], dp[n - 1][1]);
    }
};

File: Algorithm/subjects/101-200/198. 打家劫舍/main.java
class Solution {
    public int rob(int[] nums) {
        if (nums.length == 0 || nums == null) {
            return 0;
        }
        int[][] dp = new int[nums.length][2];
        //边界条件
        dp[0][0] = 0;
        dp[0][1] = nums[0];
        for (int i = 1; i < nums.length; i++) {
            //当前房子没有被偷所获得的最大金额
            dp[i][0] = Math.max(dp[i - 1][0], dp[i - 1][1]);
            //当前房子被偷了获得的最大金额
            dp[i][1] = dp[i - 1][0] + nums[i];
        }
        return Math.max(dp[nums.length - 1][0], dp[nums.length - 1][1]);
    }
}